frontend/src/container/ReduxTlmDispatcher.jsx

```diff
@@ -68,41 +71,58 @@ export class ReduxTlmDispatcher extends React.Component {
     ])
   }
 
+  handleNotification (data) {
```

Comment:
I know there isn't a clear rule for method definition, but here all other methods are defined with `= data => {`, so I would rather keep this definition flavor.

Reply:
@grignards We need to talk about this. They are not the same.

Reply:
@Skylsmoi has more knowledge than I do to talk about it, but I'd like to add that regular functions are often linked to the context that they are invoked (this), and if I'm not wrong arrow functions are pure in that sense. Here the function doesn't seems to me that it depends of the context. Can you explain to me why in this case you this a regular function it's better ?

Reply:
Because it belongs to the prototype of the class, instead of being a field of the specific instance.

Here's the difference:

```js
class C {
    method() {
        console.log(this)
    }
}
```
is translated to:

```js
function C () {}
C.prototype.method =  {
    console.log(this)
}
````

Whereas

```js
class C {
    method = () => {
        console.log(this)
    }
}
```

is translated to:

```js
function C() {
    this.method = () => {
        console.log(this)
    }
}
```
Equivalent to:

```
function C() {
    this.method = (function () {
        console.log(this)
    }).bind(this)
}
```

You can try this by running `(new C()).hasOwnProperty('method')`. In the first case, you get false, because `method` is in the prototype of the class, while in the second case, you get `true`, because `method` is not a method: it is a field of the instance that happens to be a function. 

You are right with respect to the context `this`. And our way of doing is useful for event handlers that need `this`.

I know in the Tracim frontend we widely use `method =  () => {}` but this breaks the prototypical class model of Javascript, makes proper inheritance more difficult and is probably less memory efficient. I can revert to  `method =  () => {}` but I think we really should progressively tend towards using proper Javascript methods in Tracim. I think we should limit `method =  () => {}` to event handlers that use `this`.

Reply:
Don't worry, I'm not telling you how to do it, I'm just trying to understand why, in general I know some advantages in using arrow functions, but I don't know the advantages of the regular ones. In this case, you are using `this.props` within the function, wouldn't you need `bind`?

Reply:
I understood that you asked out of curiosity (and I answered in this spirit), and I'll probably revert anyway for consistency reasons, but to answer your question: no, no need to use bind here.

```js
class C {
    f() {
        console.log(this.answer, this instanceof C)
    }
    g() {
        this.f() // is this bound here?
    }
}

const c = new C()
c.answer = 42
c.g() // prints 42 true
```

When using object.method() (or this.method()), method is bound to this object at call time.

The story is (unfortunately) different if you use a callback:

```js
function call(f) {
    f()
}

call(c.g)  // this is not bound - TypeError: Cannot read property 'f' of undefined at g, at call
```

You'd need to bind here:

```js
call(c.g.bind(c))  // prints 42 true
```

Reply:
As for an advantage of using regular methods instead of the arrow functions:
 - first, it's more concise
 - second: inheritance works:

```js
class C {
  f() {
    console.log(42)
  }
}

class D extends C {
  f () {
    super.f()
    console.log('my extension')
  }
}

d = new D()
d.f()

```
This prints:

```
42
my extension
```

Compared to:

```js
class C {
  f = () => {
    console.log(42)
  }
}

class D extends C {
  f = () => {
    super.f()
    console.log('my extension')
  }
}

const d = new D()
d.f()
```

This gives:

```
TypeError: (intermediate value).f is not a function
    at D.f (/home/raphael/t.js:9:11)
    at Object.<anonymous> (/home/raphael/t.js:15:3)
    at Module._compile (internal/modules/cjs/loader.js:955:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:991:10)
    at Module.load (internal/modules/cjs/loader.js:811:32)
    at Function.Module._load (internal/modules/cjs/loader.js:723:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1043:10)
    at internal/main/run_main_module.js:17:11
```

Classes written as we write them in Tracim cannot really be extended.

Reply:
If I may, while I agree with all the above, to me, they are not good enough reason to state a clear rule.

Here's why:

About `this` binding, there is an eslint plugin that already does it for every functions of the components.

About arrow function makings it harder for inheritance, it's true but we don't do inheritance it is complicated for components because they already all extends from React.Component. And React guidelines are to avoid inheritance.

About arrow function having poorer performances, it's also true but the loss can't make any difference on our app since we don't instantiate the same components that often on pages. While this is only true for modern browsers, we already have poor support for old ones and we barely test on them (like IE).
___
Originally, when I was the only frontend dev on Tracim and when I didn't know all the differences between the 2 function declarations, I used the regular function declaration for inherited functions from React.Component (componentDidUpdate, render ...) and everything else was arrow functions. I'm not saying it's a good rule.

I'm not against stating a clear rule but I would prefer it to be for a good reason.
Remember that, if you state a strict rule, it comes with a choice: refactor all our code base or agree that our own code base doesn't respect the rule.
Obviously, refactoring is better, but it comes with a cost.

This conversation is a good technical dept subject that can be put in an issue with the according tag. Things we, Raphael, maybe should have already done since we already discussed about it. It's just that I don't think now a good time to decide and apply said rule.

For now, the most important thing to me is knowing that all developers working on tracim know the differences between regular functions and arrow functions.

Reply:
Thanks for your comment @Skylsmoi. In any case, I don't think that we should change everything in the code base should we adopt a new rule (for anything, not necessarily on this specific point) (which is not today anyway). This is costly and error prone. But we should allow ourselves to adopt new rules for new code and gradually moving.

> For now, the most important thing to me is knowing that all developers working on tracim know the differences between regular functions and arrow functions.

Me too :-)
and class methods vs regular functions / arrow functions assigned to each instance.